src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,200 @@
 
+import java.util.Collection;
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T item, Node<T> next) {
+            this.prev = prev;
+            this.item = item;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index " + index
+                    + " out of bounce of size " + size);
+        }
+
+        if (index == size) {
+            add(value);
+        } else {
+            linkBefore(value, getNode(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        addAll(size, list);
+    }
+
+    private void addAll(int index, Collection<? extends T> collection) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index " + index
+                    + " out of bounce of size " + size);
+        }
+
+        if (collection.isEmpty()) {
+            throw new RuntimeException("Can't add elements from empty collection " + collection);
+        }
+        Object[] objects = collection.toArray();
+
+        if (index == size) {
+            for (Object o : objects) {
+                add((T) o);
+            }
+        } else {
+            for (Object o : objects) {
+                linkBefore((T) o, getNode(index));
+                index++;
+            }
+        }
+
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> setNode = getNode(index);
+        final T oldValue = setNode.item;
+        setNode.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(getNode(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> current = head; current != null; current = current.next) {
+                if (current.item == null) {
+                    unlink(current);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> current = head; current != null; current = current.next) {
+                if (object.equals(current.item)) {
+                    unlink(current);
+                    return true;
+                }
+            }
+        }
```

Comment:
make it single for loop and modify the if the condition so that it would cover both null/non-null cases
hint: check the `Objects.equals()` source code

Reply:
> `Objects.equals()` source code

done